R/data-storage.R

```diff
@@ -91,6 +98,18 @@ DataStorage <- R6::R6Class( # nolint object_name_linter
   private = list(
     .username = NULL,
     .session_id = NULL,
+    check_date = function(date_value, .var_name) {
+      if (checkmate::test_string(date_value)) {
+        date_value <- tryCatch(
+          as.Date(date_value),
+          error = function(err) {
+            date_value
+          }
+        )
+      }
+      checkmate::assert_date(date_value, .var.name = .var_name)
+      date_value
+    },
```

Comment:
Maybe we can use `tryCatch()` here more efficiently. If `date_value` is not convertible to a date, then `checkmate::assert_date()` will also throw an error. I think we can wrap both in a `tryCatch` instead and return a better error message with `logger`.

```
check_date <- function(date_value, .var_name) {
    tryCatch({
        if (checkmate::test_string(date_value))
            date_value <- as.Date(date_value)
        
        checkmate::assert_date(date_value, .var.name = .var_name)
        return(date_value)
    }, error = function(e) {
        logger::log_error(e$message)
    })
}
```
It also has a nice consequence of making sure `.var.name` is supplied.

```
[add-plumber*] 414.38 MB ❯ check_date(Sys.Date())
[1] "2023-03-29"
[add-plumber*] 414.38 MB ❯ check_date(as.Date("2023-03-29"))
[1] "2023-03-29"
[add-plumber*] 414.38 MB ❯ check_date("2023-03-29")
[1] "2023-03-29"
[add-plumber*] 414.38 MB ❯ check_date("2023-03-32")
ERROR [2023-03-29 14:22:00] character string is not in a standard unambiguous format
[add-plumber*] 414.38 MB ❯ check_date(123)
ERROR [2023-03-29 14:22:01] argument ".var.name" is missing, with no default
[add-plumber*] 414.38 MB ❯ check_date(123,.var_name = "date")
ERROR [2023-03-29 14:22:01] Assertion on 'date' failed: Must be of class 'Date', not 'double'.
[add-plumber*] 414.38 MB ❯ check_date(123, .var_name = NA)
ERROR [2023-03-29 14:22:02] Assertion on '.var.name' failed: May not be NA.
```

Reply:
That was the idea, this call must fail if the parameter is not convertible to a valid date.

But I see that it would be better to have a good error message, this is my suggestion:

```R
check_date = function(date_value, .var_name) {
      # required parameter
      checkmate::assert_string(.var_name)

      tryCatch({
        date_value <- as.Date(date_value)
        checkmate::assert_date(date_value, .var.name = .var_name)
        date_value
      }, error = function(err) {
        date_value
        rlang::abort(glue::glue(
          "Assertion on '{.var_name}' failed: Must be of class 'Date' ",
          "or a valid date format of class 'String' ('yyyy-mm-dd')."
        ))
      })
    }
```

![image](https://user-images.githubusercontent.com/211358/228460858-79a8b38f-31be-458f-a9ae-801b24bff72e.png)

Reply:
Looks good and works as expected.